docs/sc-memory/api/cpp/extended/agents/actions.md

```diff
@@ -0,0 +1,345 @@
+auto const & [argAddr1, argAddr2] = action.GetArguments<2>();
+// If action doesn't have some argument, it will be empty.
+...
+```
+
+```cpp
+...
+// You can't ignore some arguments.
+auto const & [argAddr1, argAddr2, _] = action.GetArguments<3>();
+...
+```
+
+!!! note
+    `GetArguments` uses `GetArgument` for each argument.
+
+### **SetArgument**
+
+You can set arguments for specified action.
+
+```cpp
+...
+// Provide relation here if your argument should have specific role in action. 
+action.SetArgument(ScKeynodes::rrel_1, argAddr);
+...
+```
+
+```cpp
+...
+// Or provide index of argument. 1 is equal to `ScKeynodes::rrel_1`,
+// 2 is equal to `ScKeynodes::rrel_2` and etc.
+action.SetArgument(1, argAddr);
+...
+```
+
+!!! note
+    If action has already argument with specified role, then connection between action and this argument will be removed and created new one between action and new argument.
+
+### **SetArguments**
+
+```cpp
+...
+// Set several arguments simultaneously.
+action.SetArguments(argAddr1, argAddr2);
+...
+```
+
+!!! note
+    `SetArguments` uses `SetArgument` for each argument.
+
+### **Action result**
+
+All action should have result (result situation). Result situation is structure that contains all sc-constructions that which indicate result of interpreting action.
+
+#### **GetResult**
+
+You can result for any actions.
+
+```cpp
+...
+// Use this method after that some agent finished 
+// interpreting action.
+ScStructure const & actionResult = action.GetResult();
+...
+```
+
+!!! warning
+    You can call this method for finished action only.
+
+#### **SetResult**
+
+```cpp
+...
+// Use this method in agent interpreting action 
+// to set new result.
+action.SetResult(resultStructure);
+...
+```
+
+!!! note
+    If action has result, then will be removed and the new one will be set.
+
+!!! warning
+    You can call this method for not finished, but initiated action only.
+
+#### **FormResult**
+
+You must not create result structure. You can provide only elements of result for action.
+
+```cpp
+...
+// Use this method in agent interpreting action 
+// to form new result with provided sc-elements.
+action.FormResult(elementAddr1, elementAddr2);
+...
+```
+
+!!! note
+    If action has result, then will be removed and the new one will be set.
+
+#### **UpdateResult**
+
+```cpp
+...
+// Use this method in agent interpreting action
+// for update existing result by new sc-elements.
+action.UpdateResult(elementAddr1, elementAddr2);
+...
+```
+
+!!! note
+    This method updates existing result for action.
+
+!!! note
+    `FormResult` and `UpdateResult` doesn't append sc-element twice.
+
+!!! note
+    If you don't form result then empty result for your action will be generated if you finish action.
+
+### **Action states**
+
+All actions have state. There are three states of actions provided by this API:
+
+* action isn't initiated;
+* action is initiated, but isn't finished, that is, action is performed;
```

Comment:
```suggestion
* action is initiated, but isn't finished, that is, action is being performed;
```
also I think that it is possible that action is initiated but it wasn't started because it is still in the events queue(only one thread that processes events and a lot of people are using sc-web)

Reply:
This is true, but it describes the types of actions according to what methods this API provides.

Reply:
why is this API returning results that aren't true

Reply:
IsInitiated() && !IsFinished()

Reply:
this file introduces action state 'performed' but it is never used as action state in this file, I don't understand why this is needed here at all